src/Microsoft.AspNetCore.B2CIntegration/AzureAdB2COpenIDConnectEventHandlers.cs

```diff
@@ -0,0 +1,88 @@
+            if (context.Properties.Items.TryGetValue(AzureAdB2CDefaults.PolicyKey, out var policy) &&
+                !string.IsNullOrEmpty(policy) &&
+                !string.Equals(policy, defaultPolicy, StringComparison.OrdinalIgnoreCase))
+            {
+                context.ProtocolMessage.Scope = OpenIdConnectScope.OpenIdProfile;
+                context.ProtocolMessage.ResponseType = OpenIdConnectResponseType.IdToken;
+                context.ProtocolMessage.IssuerAddress = BuildIssuerAddress(context, defaultPolicy, policy);
+                context.Properties.Items.Remove(AzureAdB2CDefaults.PolicyKey);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private string BuildIssuerAddress(RedirectContext context, string defaultPolicy, string policy)
+        {
+            if (!_policyToIssuerAddress.TryGetValue(policy, out var issuerAddress))
+            {
+                _policyToIssuerAddress[policy] = context.ProtocolMessage.IssuerAddress.ToLowerInvariant()
+                    .Replace($"/{defaultPolicy.ToLowerInvariant()}/", $"/{policy.ToLowerInvariant()}/");
+            }
+
+            return _policyToIssuerAddress[policy];
+        }
+
+        public Task OnRemoteFailure(RemoteFailureContext context)
+        {
+            context.HandleResponse();
+            // Handle the error code that Azure AD B2C throws when trying to reset a password from the login page 
+            // because password reset is not supported by a "sign-up or sign-in policy".
+            // Below is a sample error message:
+            // 'access_denied', error_description: 'AADB2C90118: The user has forgotten their password.
+            // Correlation ID: f99deff4-f43b-43cc-b4e7-36141dbaf0a0
+            // Timestamp: 2018-03-05 02:49:35Z
+            //', error_uri: 'error_uri is null'.
+            if (context.Failure is OpenIdConnectProtocolException && context.Failure.Message.Contains("AADB2C90118"))
+            {
+                // If the user clicked the reset password link, redirect to the reset password route
+                context.Response.Redirect($"/Account/ResetPassword/{SchemeName}");
+            }
+            // Access denied errors happen when a user cancels an action on the Azure AD B2C UI. We just redirect back to
+            // the main page in that case.
+            // Message contains error: 'access_denied', error_description: 'AADB2C90091: The user has cancelled entering self-asserted information.
+            // Correlation ID: d01c8878-0732-4eb2-beb8-da82a57432e0
+            // Timestamp: 2018-03-05 02:56:49Z
+            // ', error_uri: 'error_uri is null'.
+            else if (context.Failure is OpenIdConnectProtocolException && context.Failure.Message.Contains("access_denied"))
+            {
+                context.Response.Redirect("/");
+            }
+            else
+            {
+                context.Response.Redirect("/Error");
```

Comment:
I'm thinking about adding an error page to make this UI self contained (and possibly move everything to its own area so that it doesn't clash with anything else)